cpp/src/arrow/array.h

```diff
@@ -203,12 +224,14 @@ struct ARROW_EXPORT ArrayData {
   // Construct a zero-copy slice of the data with the indicated offset and length
   ArrayData Slice(int64_t offset, int64_t length) const;
 
+  void SetNullCount(int64_t v) { null_count.store(v); }
+
   /// \brief Return null count, or compute and set it if it's not known
   int64_t GetNullCount() const;
 
   std::shared_ptr<DataType> type;
   int64_t length;
-  mutable int64_t null_count;
+  mutable std::atomic<int64_t> null_count;
```

Comment:
Implicit conversion is nice but if the intention is to be thread safe then it seems we should make `ArrayData::null_count` private and only access it through `ArrayData::{Get,Set}NullCount`. Is there some use case for a public atomic I'm not seeing (like CES of null_count in some shared ArrayData maybe)? `ArrayData` is an internal structure so I wouldn't block on this in any case

Reply:
Indeed no code (except unit tests) should be using `null_count` directly (since it may be uncomputed). This is an issue with the existing code already -- if we want to make this member private we can do that in a follow-up PR since it would probably be more work